### PR TITLE
Add normisedPackageIdExists predicate, and use it in upload

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by phadej** [comment]
Ah, No upper-bounds strikes again :S

File: Distribution/Server/Features/Core.hs [hunk 1]
```diff
@@ -715,3 +716,21 @@ packageExists, packageIdExists :: (Package pkg, Package pkg') => PackageIndex pk
 packageExists   pkgs pkg = not . null $ PackageIndex.lookupPackageName pkgs (packageName pkg)
 -- | Whether a particular package version exists in the given package index.
 packageIdExists pkgs pkg = maybe False (const True) $ PackageIndex.lookupPackageId pkgs (packageId pkg)
+
+-- | Whether a parituclar normalised version exists in the given package index.
```
Comment (hvr) [inline Distribution/Server/Features/Core.hs]: typo in comment

File: Distribution/Server/Features/Core.hs [hunk 2]
```diff
@@ -715,3 +716,21 @@ packageExists, packageIdExists :: (Package pkg, Package pkg') => PackageIndex pk
...
+    elem (normalisedPackageId pkg) $ map normalisedPackageId $ PackageIndex.lookupPackageName pkgs (packageName pkg)
+  where
+    normalisedPackageId :: Package pkg  => pkg -> PackageIdentifier
+    normalisedPackageId pkg' = case packageId pkg' of
+        PackageIdentifier name ver -> PackageIdentifier name (normaliseVersion ver)
+
+    normaliseVersion :: Version -> Version
+    normaliseVersion (Version vs _) = Version (n vs) []
+      where
+        n vs' = case foldr f [] vs' of
```
Comment (hvr) [inline Distribution/Server/Features/Core.hs]: with a recent enough `base`, we could use `dropWhileEnd` :-)

**Comment by phadej** [comment]
fixed, seems we are `base >= 4.6`, so `dropWhileEnd` is there.

**Comment by hvr** [comment]
@cartazio @bergmark @dcoutts As this effectively imposes a new Hackage policy, I'd like you to comment on whether you agree with this new check, which will prohibit uploading new package versions to Hackage which only differ by the amount of trailing `0` version components from existing package versions.

**Comment by bergmark** [comment]
I agree, as with case sensitive package names the current situation can be very confusing.

**Comment by cartazio** [comment]
What preexisting code bases / packages  on hackage violate this new rule? Would the cabal solver interpret those correctly ?

Naively, I do agree with bergmark, and am likewise supportive of trailing zeros being an equivalence class.  But might be good to understand impact.  

Speaking of, what's our currency policy on case?

**Comment by phadej** [comment]
@cartazio see https://github.com/haskell/cabal/pull/3515#issuecomment-231714012

cabal isn't affected by this at all.

**Comment by dcoutts** [comment]
What is the motivation for this policy? The version ordering is quite clear internally, it's lexicographic. Is this for the humans or the machines?

**Comment by phadej** [comment]
@dcoutts, for humans. This doesn't change the ordering, if you have `package-1` and `package-1.0` in the index, they are still there and separate.

This only prevents uploading new versions of packages, which differ from existing ones by amount of trailing zeros.

> Let me even take a step back and ask: is there some de facto interpretation of the PVP which conflicts with the letter of the PVP? Perhaps on the relation of 1.0 and 1? If so, maybe the letter of the PVP should change. I've always thought that 1.0 and 1.0.0 are equivalent; if this were not the case, that's surprising!

PVP isn't clear about this. IMHO from three options
- make `1.0` and `1.0.0` equivalent
- prevent new `1.0` and `1.0.0` -like uploads
- change PVP wording

the second one is most conservative

---

Edward said in https://github.com/haskell/cabal/pull/3515#issuecomment-231586996

> There are maybe a half dozen packages that have both v1 and v1.0 versions, and nobody cares. We could fix the version check to effectively zero pad, delete the conflicting versions from hackage, make it so it doesn't accept new instances of this problem and not be stuck explaining this for decades to come.
>
> We're struggling to find a path of least breakage for a misfeature.

I'm not sure about deleting, but preventing new instances of the problem is :+1:

---

In practice this isn't a huge problem, as there aren't many instances. But atm `MIN_VERSION_package` cannot distinguish between them, and you cannot fix `MIN_VERSION_` as it will break almost every ekmett's package. 

That's why I made https://github.com/haskell/cabal/pull/3515#issuecomment-231714012, but this is much more pragmatic.

**Comment by dcoutts** [comment]
Merged with a couple tweaks.

**Pull request closed without merging**